### PR TITLE
Add support for using pycoap for communicating with the gateway

### PR DESCRIPTION
**Pull request by moroen**
Added API to use pycoap [](https://github.com/moroen/python-coap-module) as an alternative to coap-client for synchronous communication with the IKEA gateway.

This is intended as a drop-in alternative to coap-client, all existing code should work unchanged except importing the APIFactory:

`from pytradfri.api.pycoap_api import APIFactory`

Admittedly this change is tailored for the use of pytradfri in the IKEA-Tradfri plugin for Domoticz, but hopefully provide some value in other projects as well.

Advantages over coap-client:
- Native python extension module
- Only two dependencies for installation: openssl@1.1 and golang compiler
- Compiles and installs using standard python setuptools

Limitations over coap-client:
- Only works on linux and macOS-versions prior to Mojave (High Sierra and downwards)
- Currently doesn't suppport observations - support for this is planned

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
+cython
```
Comment (ggravlingen) [inline requirements.txt]: Please pin the versoin

**Comment by ggravlingen** [comment]
I don't mind adding support for pycoap, but reckon we should add an extras require to `setup.py`.
```
EXTRAS_REQUIRE = {
    'pycoap': ['cython?', 'golang?', 'pycoap?']
}
```

Also, we should update the readme to indicate support is less developed for pycoap.

File: requirements.txt (hunk 1 above)
Comment (ggravlingen) [reply requirements.txt]: Oh, and isn't pycoap a requirement?

**Comment by moroen** [comment]
Thank you for the feedback, and thank you for being interested in adding support pycoap. 

I'm working on getting pycoap to work with manylinux in order to publish it on pypi, which  I guess is an requirement for installing it via EXTRAS_REQUIRE in setup.py. As soon as I have this sorted, I'll update the pull request including the required changes in setup.py and readme.

**Comment by ggravlingen** [comment]
Sounds good! Please also add some tests for the api, we want to make sure things are not broken if others contribute to the code.

**Comment by moroen** [comment]
Getting the extension compiled on CentOS 5 for manylinux wheel is turning out to be a total dependency nightmare, I'm ending up with upgrading so many components of centos in the manylinux docker image, that for all practical purposes, it's no longer centos 5...

So I wonder, how much of a problem is it to require pycoap to be installed manually using 
`python3 setup.py install` 

I would argue that it's no more a hassle than compiling coap-client manually, which kind of was the intention...

**Comment by ggravlingen** [comment]
You mean fully leave it to the user to make sure pycoap is working before they install pytradfri?

**Comment by moroen** [comment]
Upgrading to the latest version of openssl has made this extension unworkable in unexpected ways, so at the moment it's "dead in the water". Thank you for your interest in allowing this, I'm closing the pull request, and issue a new one when I have solved the current problem with openssl.

**Pull request closed without merging**